### ISSUE TITLE
etcd-wrapper with etc-etcd volume

### ISSUE THREAD

**Issue by urzds** [coreos/bugs#1799]
# Issue Report #

### Environment ###

What hardware/cloud provider/hypervisor is being used to run Container Linux?

bare-metal

### Desired Feature ###

It would be nice if `/usr/lib/coreos/etcd-wrapper` would use a `etc-etcd` volume, just like `/usr/lib/coreos/kubelet-wrapper` uses a `etc-kubernetes` volume. This would make it easier to pass custom files into the container.

Previously, with `etcd2.service`, I stored my SSL certificates at `/etc/etcd/ssl`. With `etcd-member.service` I have to store them in `/var/lib/etcd/ssl`, because that is the only etcd-specific directory that is passed into the container. This seems to be the wrong location, because etcd complains: `etcdmain: found invalid file/dir ssl under data dir /var/lib/etcd (Ignore this if you are upgrading etcd)`

### Other Information ###

CoreOS Container Linux 1235.9.0, etcd 3.1.0

**Comment by crawford** [comment]
You can set `ETCD_SSL_DIR` in a drop-in to the wrapper. This will be mounted to `/etc/ssl/certs` inside the container.

**Comment by urzds** [comment]
But `/etc/ssl/certs` is usually used for the hashes of certificates from `/usr/share/ca-certificates`, is it not?

**Comment by crawford** [comment]
It is, but etcd shouldn't use the root certificates, so it should be safe just to add your certificates.

**Comment by euank** [comment]
Please feel free to re-open if @crawford's answer / `ETCD_SSL_DIR` doesn't resolve this.

Closing for now.

**Comment by emmanuel** [comment]
In general I'm elated that there is already a facility for overriding the host path that will be mounted into the etcd containers at `/etc/ssh/certs`. Unfortunately, I've just attempted to utilize this, and discovered that I am unable to verify the certificate from `https://discovery.etcd.io`. I get the point that etcd should *not* trust everything in the OS trust store, but it is inconvenient to merge in the CA for `discovery.etcd.io` with my internal CA to support both in one installation. One on level, I suppose this is merely a complaint about the effort ;). On another level, I suppose this is a question: does anyone on this issue have suggestions about how to easily support an internal CA along (for etcd member comms) with an external CA (for verifying discovery).

**Comment by urzds** [comment]
As I said, I store my own CA and certificates in `/var/lib/etcd/ssl` (which makes etcd complain about files of unknown origin, but I just ignore that warning) and have the system CAs in `/etc/ssl/certs`. etcd is then started with `--ca-file=/var/lib/etcd/ssl/...` and that does the trick for me.

If you need both the system CAs and your own CA in one file, the way Linux distributions handle this in general is by letting the user place his CAs in `/usr/local/share/ca-certificates`, while they place the system CAs in `/usr/share/ca-certificates` and then `update-ca-certificates` generates `/etc/ssl/certs/ca-certificates.crt` from all files in both locations. Maybe you can build something along these lines.

**Comment by emmanuel** [comment]
Right, that makes sense. I was missing the part about the internal CA piggy-backing on the mount for the data volume. In my question I omitted to mention that I started by trying to shave the minimal effort of adding a volume and mount to the `RKT_OPTS` env var from a drop-in to `etcd-member`. 

I like the idea of etcd only trusting the CAs explicitly needed (i.e. not trusting `/etc/ssl/certs/ca-certificates.crt`), so I added an `ExecStartPre` to a drop-in to `etcd-member` that verifies https://discovery.etcd.io then fetches the cert chain from it and then starts up etcd with only that chain in the container's `/etc/ssl/certificates/ca-certificates.crt`. 

In other words, thanks! Before this thread I didn't even conceive of removing unneeded CAs from a 'system' container (kubelet and now etcd, under rkt).

**Comment by euank** [comment]
Ah, thanks for the reminder discovery needs certs too!

Reopening since it sounds like this should be improved.
